Core/UnModified_OpenSource/autogen/autogen-5.18.12/agen5/autogen.1
.de1 NOP
.  it 1 an-trap
.  if \\n[.$] \,\\$*\/
..
.ie t \
.ds B-Font [CB]
.ds I-Font [CI]
.ds R-Font [CR]
.el \
.ds B-Font B
.ds I-Font I
.ds R-Font R
.TH autogen 1 "29 Aug 2016" "GNU AutoGen (5.18.12)" "User Commands"
.\"
.\" DO NOT EDIT THIS FILE (in-mem file)
.\"
.\" It has been AutoGen-ed
.\" From the definitions /u/bkorb/ag/ag/agen5/opts.def
.\" and the template file agman-cmd
.SH NAME
\f\*[B-Font]autogen\fP
\- The Automated Program Generator
.SH SYNOPSIS
\f\*[B-Font]autogen\fP
.\" Mixture of short (flag) options and long options
[\f\*[B-Font]\-flags\f[]]
[\f\*[B-Font]\-flag\f[] [\f\*[I-Font]value\f[]]]
[\f\*[B-Font]\-\-option-name\f[][[=| ]\f\*[I-Font]value\f[]]]
[ <def-file> ]
.sp \n(Ppu
.ne 2

AutoGen creates text files from templates using external definitions.
.SH DESCRIPTION
\fBAutoGen\fP is designed for generating program files that contain
repetitive text with varied substitutions.  The goal is to simplify the
maintenance of programs that contain large amounts of repetitious text.
This is especially valuable if there are several blocks of such text
that must be kept synchronized.
.sp
One common example is the problem of maintaining the code required for
processing program options.  Processing options requires a minimum of
four different constructs be kept in proper order in different places
in your program.  You need at least: The flag character in the flag
string, code to process the flag when it is encountered, a global
state variable or two, and a line in the usage text.
You will need more things besides this if you choose to implement
long option names, configuration file processing, environment variables
and so on.
.sp
All of this can be done mechanically; with the proper templates
and this program.
.SH "OPTIONS"
.SS "The following options select definitions, templates and scheme functions to use"
.TP
.NOP \f\*[B-Font]\-L\f[] \f\*[I-Font]dir\f[], \f\*[B-Font]\-\-templ\-dirs\f[]=\f\*[I-Font]dir\f[]
Search for templates in \fIDIR\fP.
This option may appear an unlimited number of times.
.sp
Add a directory to the list of directories \fBautogen\fP searches when
opening a template, either as the primary template or an included one.
The last entry has the highest priority in the search list.  That is
to say, they are searched in reverse order.
.TP
.NOP \f\*[B-Font]\-T\f[] \f\*[I-Font]tpl\-file\f[], \f\*[B-Font]\-\-override\-tpl\f[]=\f\*[I-Font]tpl\-file\f[]
Use \fITPL-FILE\fP for the template.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
Definition files specify the standard template that is to be expanded.
This option will override that name and expand a different template.
.TP
.NOP \f\*[B-Font]\-l\f[] \f\*[I-Font]tpl\-file\f[], \f\*[B-Font]\-\-lib\-template\f[]=\f\*[I-Font]tpl\-file\f[]
Load AutoGen macros from \fITPL-FILE\fP.
This option may appear an unlimited number of times.
.sp
DEFINE macros are saved from this template file for use in processing
the main macro file.  Template text aside from the DEFINE macros is
is ignored.
.sp
Do not use this.  Instead, use the INCLUDE macro in your template.
.sp
.B
NOTE: THIS OPTION IS DEPRECATED
.TP
.NOP \f\*[B-Font]\-\-definitions\f[]=\f\*[I-Font]file\f[], \f\*[B-Font]\-\-no\-definitions\f[]
Read definitions from \fIFILE\fP.
The \fIno\-definitions\fP form will disable the option.
This option is enabled by default.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
Use this argument to specify the input definitions file with a
command line option.  If you do not specify this option, then
there must be a command line argument that specifies the file,
even if only to specify stdin with a hyphen (\fB-\fP).
Specify, \fB--no-definitions\fP when you wish to process
a template without any active AutoGen definitions.
.TP
.NOP \f\*[B-Font]\-\-shell\f[]=\f\*[I-Font]shell\f[]
name or path name of shell to use.
.sp
By default, when AutoGen is built, the configuration is probed for a
reasonable Bourne-like shell to use for shell script processing.  If
a particular template needs an alternate shell, it must be specified
with this option on the command line, with an environment variable
(\fBSHELL\fP) or in the configuration/initialization file.
.TP
.NOP \f\*[B-Font]\-m\f[], \f\*[B-Font]\-\-no\-fmemopen\f[]
Do not use in-mem streams.
.sp
If the local C library supports "\fBfopencookie(3GNU)\fP", or
"\fBfunopen(3BSD)\fP" then AutoGen prefers to use in-memory stream
buffer opens instead of anonymous files.  This may lead to problems
if there is a shortage of virtual memory.  If, for a particular
application, you run out of memory, then specify this option.
This is unlikely in a modern 64-bit virtual memory environment.
.sp
On platforms without these functions, the option is accepted
but ignored.  \fBfmemopen(POSIX)\fP is not adequate because
its string buffer is not reallocatable.  \fBopen_memstream(POSIX)\fP
is \fIalso\fP not adequate because the stream is only opened for
output.  AutoGen needs a reallocatable buffer available for both
reading and writing.
.TP
.NOP \f\*[B-Font]\-\-equate\f[]=\f\*[I-Font]char\-list\f[]
characters considered equivalent.
The default
\f\*[I-Font]char\-list\f[]
for this option is:
.ti +4
 _-^
.sp
This option will alter the list of characters considered equivalent.
The default are the three characters, "_-^".  (The last is conventional
on a Tandem/HP-NonStop, and I used to do a lot of work on Tandems.)
.SS "The following options modify how output is handled"
.TP
.NOP \f\*[B-Font]\-b\f[] \f\*[I-Font]name\f[], \f\*[B-Font]\-\-base\-name\f[]=\f\*[I-Font]name\f[]
Specify \fBNAME\fP as the base name for output.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
A template may specify the exact name of the output file.  Normally,
it does not.  Instead, the name is composed of the base name of the
definitions file with suffixes appended.  This option will override the
base name derived from the definitions file name.  This is required if
there is no definitions file and advisable if definitions are being
read from stdin.  If the definitions are being read from standard in,
the base name defaults to \fIstdin\fP.  Any leading directory components
in the name will be silently removed.  If you wish the output file to
appear in a particular directory, it is recommended that you "cd" into
that directory first, or use directory names in the format specification
for the output suffix lists, see: pseudo macro.
.TP
.NOP \f\*[B-Font]\-\-source\-time\f[], \f\*[B-Font]\-\-no\-source\-time\f[]
set mod times to latest source.
The \fIno\-source\-time\fP form will disable the option.
.sp
If you stamp your output files with the \fBDNE\fP macro output, then
your output files will always be different, even if the content has
not really changed.  If you use this option, then the modification
time of the output files will change only if the input files change.
This will help reduce unneeded builds.
.TP
.NOP \f\*[B-Font]\-\-writable\f[], \f\*[B-Font]\-\-not\-writable\f[]
Allow output files to be writable.
The \fInot\-writable\fP form will disable the option.
.sp
This option will leave output files writable.
Normally, output files are read-only.
.SS "The following options are often useful while debugging new templates"
They specify limits that prevent the template from taking overly long
or producing more output than expected.
.TP
.NOP \f\*[B-Font]\-\-loop\-limit\f[]=\f\*[I-Font]lim\f[]
Limit on increment loops.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]lim\f[]
is constrained to being:
.in +4
.nf
.na
exactly \-1, or
in the range  1 through 0x1000000
.fi
.in -4
The default
\f\*[I-Font]lim\f[]
for this option is:
.ti +4
 256
.sp
This option prevents runaway loops.  For example, if you accidentally
specify, "FOR x (for-from 1) (for-to \-1) (for-by 1)", it will take a
long time to finish.  If you do have more than 256 entries in tables,
you will need to specify a new limit with this option.
.TP
.NOP \f\*[B-Font]\-t\f[] \f\*[I-Font]seconds\f[], \f\*[B-Font]\-\-timeout\f[]=\f\*[I-Font]seconds\f[]
Limit server shell operations to \fBSECONDS\fP.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]seconds\f[]
is constrained to being:
.in +4
.nf
.na
in the range  0 through 3600
.fi
.in -4
.sp
AutoGen works with a shell server process.  Most normal commands will
complete in less than 10 seconds.  If, however, your commands need more
time than this, use this option.
.sp
The valid range is 0 to 3600 seconds (1 hour).
Zero will disable the server time limit.
.TP
.NOP \f\*[B-Font]\-\-trace\f[]=\f\*[I-Font]level\f[]
tracing level of detail.
This option takes a keyword as its argument.  The argument sets an enumeration value that can
be tested by comparing them against the option value macro.
The available keywords are:
.in +4
.nf
.na
nothing       debug-message server-shell
templates     block-macros  expressions
everything
.fi
or their numeric equivalent.
.in -4
.sp
The default
\f\*[I-Font]level\f[]
for this option is:
.ti +4
 nothing
.sp
This option will cause AutoGen to display a trace of its template
processing.  There are six levels, each level including messages from
the previous levels:
.sp
.sp
.IR "nothing"
Does no tracing at all (default)
.sp
.sp
.IR "debug-message"
Print messages from the "DEBUG" AutoGen macro (see: DEBUG).
.sp
.sp
.IR "server-shell"
Traces all input and output to the server shell.  This includes a shell
"independent" initialization script about 30 lines long.  Its output is
discarded and not inserted into any template.
.sp
.sp
.IR "templates"
Traces the invocation of \fBDEFINE\fPd macros and \fBINCLUDE\fPs
.sp
.sp
.IR "block-macros"
Traces all block macros.  The above, plus \fBIF\fP, \fBFOR\fP,
\fBCASE\fP and \fBWHILE\fP.
.sp
.sp
.IR "expressions"
Displays the results of expression evaluations.
.sp
.sp
.IR "everything"
Displays the invocation of every AutoGen macro, even \fBTEXT\fP macros
(i.e. the text outside of macro quotes).  Additionally, if you rebuild
the ``expr.ini'' file with debugging enabled, then all calls to
AutoGen defined scheme functions will also get logged:
.br
.nf
    cd ${top_builddir}/agen5
    DEBUG_ENABLED=true bash bootstrap.dir expr.ini
    make CFLAGS='-g \-DDEBUG_ENABLED=1'
.fi
.sp
Be aware that you cannot rebuild this source in this way without first
having installed the \fBautogen\fP executable in your search path.
Because of this, "expr.ini" is in the distributed source list, and
not in the dependencies.
.br
.TP
.NOP \f\*[B-Font]\-\-trace\-out\f[]=\f\*[I-Font]file\f[]
tracing output file or filter.
.sp
The output specified may be a file name, a file that is appended to,
or, if the option argument begins with the \fBpipe\fP operator
(\fB|\fP), a command that will receive the tracing output as standard
in.  For example, \fB--traceout='| less'\fP will run the trace output
through the \fBless\fP program.  Appending to a file is specified by
preceding the file name with two greater-than characters (\fB>>\fP).
.TP
.NOP \f\*[B-Font]\-\-show\-defs\f[]
Show the definition tree.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
This will print out the complete definition tree before processing
the template.
.TP
.NOP \f\*[B-Font]\-\-used\-defines\f[]
Show the definitions used.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
This will print out the names of definition values searched for
during the processing of the template, whether actually found or
not.  There may be other referenced definitions in a template in
portions of the template not evaluated.  Some of the names listed
may be computed names and others AutoGen macro arguments.  This is
not a means for producing a definitive, all-encompassing list of all
and only the values used from a definition file.  This is intended
as an aid to template documentation only.
.TP
.NOP \f\*[B-Font]\-C\f[], \f\*[B-Font]\-\-core\f[]
Leave a core dump on a failure exit.
.sp
Many systems default to a zero sized core limit.  If the system
has the sys/resource.h header and if this option is supplied,
then in the failure exit path, autogen will attempt to set the
soft core limit to whatever the hard core limit is.  If that
does not work, then an administrator must raise the hard core
size limit.
.Ss "These options can be used to control what gets processed
in the definitions files and template files"
They specify which outputs and parts of outputs to produce.
.TP
.NOP \f\*[B-Font]\-s\f[] \f\*[I-Font]suffix\f[], \f\*[B-Font]\-\-skip\-suffix\f[]=\f\*[I-Font]suffix\f[]
Skip the file with this \fISUFFIX\fP.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
This option must not appear in combination with any of the following options:
select-suffix.
.sp
Occasionally, it may not be desirable to produce all of the output
files specified in the template.  (For example, only the \fI.h\fP
header file, but not the \fI.c\fP program text.)  To do this
specify \fB--skip-suffix=c\fP on the command line.
.TP
.NOP \f\*[B-Font]\-o\f[] \f\*[I-Font]suffix\f[], \f\*[B-Font]\-\-select\-suffix\f[]=\f\*[I-Font]suffix\f[]
specify this output suffix.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
If you wish to override the suffix specifications in the template,
you can use one or more copies of this option.  See the suffix
specification in the @ref{pseudo macro} section of the info doc.
.TP
.NOP \f\*[B-Font]\-D\f[] \f\*[I-Font]value\f[], \f\*[B-Font]\-\-define\f[]=\f\*[I-Font]value\f[]
name to add to definition list.
This option may appear an unlimited number of times.
.sp
The AutoGen define names are used for the following purposes:
.sp
.sp 1
Sections of the AutoGen definitions may be enabled or disabled
by using C-style #ifdef and #ifndef directives.
.sp 1
When defining a value for a name, you may specify the index
for a particular value.  That index may be a literal value,
a define option or a value #define-d in the definitions themselves.
.sp 1
The name of a file may be prefixed with \fB$NAME/\fP.
The \fB$NAME\fP part of the name string will be replaced with
the define-d value for \fBNAME\fP.
.sp 1
When AutoGen is finished loading the definitions, the defined values
are exported to the environment with, \fBputenv(3)\fP.
These values can then be used in shell scripts with \fB${NAME@\fP}
references and in templates with \fB(getenv "NAME")\fP.
.sp 1
While processing a template, you may specify an index to retrieve
a specific value.  That index may also be a define-d value.
.br
.sp
It is entirely equivalent to place this name in the exported environment.
Internally, that is what AutoGen actually does with this option.
.TP
.NOP \f\*[B-Font]\-U\f[] \f\*[I-Font]name\-pat\f[], \f\*[B-Font]\-\-undefine\f[]=\f\*[I-Font]name\-pat\f[]
definition list removal pattern.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
Similar to 'C', AutoGen uses \fB#ifdef/#ifndef\fP preprocessing
directives.  This option will cause the matching names to be
removed from the list of defined values.
.SS "This option is used to automate dependency tracking"
.TP
.NOP \f\*[B-Font]\-M\f[] \f\*[I-Font]type\f[], \f\*[B-Font]\-\-make\-dep\f[] [\f\*[I-Font]type\f[]]
emit make dependency file.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
This option behaves fairly closely to the way the \fB-M\fP series of
options work with the gcc compiler, except that instead of just
emitting the predecessor dependencies, this also emits the successor
dependencies (output target files).  By default, the output dependency
information will be placed in \fB<base-name>.d\fP, but may also be
specified with \fB-MF<file>\fP.  The time stamp on this file will be
manipulated so that it will be one second older than the oldest
primary output file.
.sp
The target in this dependency file will normally be the dependency
file name, but may also be overridden with \fB-MT<targ-name>\fP.
AutoGen will not alter the contents of that file, but it may create
it and it will adjust the modification time to match the start time.
.sp
\fBNB:\fP these second letters are part of the option argument, so
\fB-MF <file>\fP must have the space character quoted or omitted, and
\fB-M "F <file>"\fP is acceptable because the \fBF\fP is part of the
option argument.
.sp
\fB-M\fP may be followed by any of the letters M, F, P, T, Q, D, or G.
However, only F, Q, T and P are meaningful.  All but F have somewhat
different meanings.  \fB-MT<name>\fP is interpreted as meaning
\fB<name>\fP is a sentinel file that will depend on all inputs
(templates and definition files) and all the output files will depend
on this sentinel file.  It is suitable for use as a real make target.
Q is treated identically to T, except dollar characters ('$') are
doubled.  P causes a special clean (clobber) phoney rule to be inserted

Core/UnModified_OpenSource/autogen/autogen-5.18.12/agen5/autogen.c

/**
 *  @file autogen.c
 *
 *  This is the main routine for autogen.
 *  @addtogroup autogen
 *  @{
 */

/*  This file is part of AutoGen.
 *  Copyright (C) 1992-2016 Bruce Korb - all rights reserved
 *
 * AutoGen is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AutoGen is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#ifdef HAVE_SYS_RESOURCE_H
# include <sys/resource.h>
#endif

#ifdef SIGRTMIN
# define MAX_AG_SIG SIGRTMIN-1
#else
# define MAX_AG_SIG NSIG
#endif
#ifndef SIGCHLD
#  define SIGCHLD SIGCLD
#endif

typedef void (void_main_proc_t)(int, char **);

typedef enum {
    EXIT_PCLOSE_WAIT,
    EXIT_PCLOSE_NOWAIT
} wait_for_pclose_enum_t;

#define _State_(n)  #n,
static char const * const state_names[] = { STATE_TABLE };
#undef _State_

typedef void (sighandler_proc_t)(int sig);
static sighandler_proc_t ignore_signal, catch_sig_and_bail;

/* = = = START-STATIC-FORWARD = = = */
static void
inner_main(void * closure, int argc, char ** argv);

static void
exit_cleanup(wait_for_pclose_enum_t cl_wait);

static _Noreturn void
cleanup_and_abort(int sig);

static void
catch_sig_and_bail(int sig);

static void
ignore_signal(int sig);

static void
done_check(void);

static void
setup_signals(sighandler_proc_t * hdl_chld,
              sighandler_proc_t * hdl_abrt,
              sighandler_proc_t * hdl_dflt);
/* = = = END-STATIC-FORWARD = = = */

#ifndef HAVE_CHMOD
#  include "compat/chmod.c"
#endif

/**
 * main routine under Guile guidance
 *
 * @param  closure Guile closure parameter.  Not used.
 * @param  argc    argument count
 * @param  argv    argument vector
 */
static void
inner_main(void * closure, int argc, char ** argv)
{
    atexit(done_check);
    initialize(argc, argv);

    processing_state = PROC_STATE_LOAD_DEFS;
    read_defs();

    /*
     *  Activate the AutoGen specific Scheme functions.
     *  Then load, process and unload the main template
     */
    processing_state = PROC_STATE_LOAD_TPL;

    {
        templ_t * tpl = tpl_load(tpl_fname, NULL);

        processing_state = PROC_STATE_EMITTING;
        process_tpl(tpl);

        processing_state = PROC_STATE_CLEANUP;
        cleanup(tpl);
    }

    processing_state = PROC_STATE_DONE;
    setup_signals(SIG_DFL, SIG_IGN, SIG_DFL);
    ag_exit_code = AUTOGEN_EXIT_SUCCESS;
    done_check();
    /* NOTREACHED */
    (void)closure;
}

/**
 * main() called from _start()
 *
 * @param  argc    argument count
 * @param  argv    argument vector
 * @return nothing -- Guile never returns, but calls exit(2).
 */
int
main(int argc, char ** argv)
{
    {
        char const * lc = getenv("LC_ALL");
        if ((lc != NULL) && (*lc != NUL) && (strcmp(lc, "C") != 0)) {
            static char lc_all[] = "LC_ALL=C";
            putenv(lc_all);
            execvp(argv[0], argv);
            fserr(AUTOGEN_EXIT_FS_ERROR, "execvp", argv[0]);
        }
    }

    // setlocale(LC_ALL, "");
    setup_signals(ignore_signal, SIG_DFL, catch_sig_and_bail);
    optionSaveState(&autogenOptions);
    trace_fp = stderr;
    prep_env();

    AG_SCM_BOOT_GUILE(argc, argv, inner_main);

    /* NOTREACHED */
    return EXIT_FAILURE;
}

/**
 * This code must run regardless of which exit path is taken
 *
 * @param [in] cl_wait Whether or not a child process should be waited for.
 */
static void
exit_cleanup(wait_for_pclose_enum_t cl_wait)
{
    /*
     *  There are contexts wherein this function can get run twice.
     */
    {
        static int exit_cleanup_done = 0;
        if (exit_cleanup_done) {
            if (OPT_VALUE_TRACE > TRACE_NOTHING)
                fprintf(trace_fp, EXIT_CLEANUP_MULLIGAN);
            return;
        }

        exit_cleanup_done = 1;
    }

#ifdef SHELL_ENABLED
    SCM_EVAL_CONST(EXIT_CLEANUP_STR);
#endif

    close_server_shell();

    if (OPT_VALUE_TRACE > TRACE_NOTHING)
        fprintf(trace_fp, EXIT_CLEANUP_DONE_FMT,
                (cl_wait == EXIT_PCLOSE_WAIT)
                ? EXIT_CLEANUP_WAITED : EXIT_CLEANUP_NOWAIT,
                (unsigned int)getpid());

    do  {
        if (trace_fp == stderr)
            break;

        if (! trace_is_to_pipe) {
            fclose(trace_fp);
            break;
        }


        fflush(trace_fp);
        pclose(trace_fp);
        if (cl_wait == EXIT_PCLOSE_WAIT) {
            int status;
            while (wait(&status) > 0)  ;
        }
    } while (false);

    trace_fp = stderr;
    fflush(stdout);
    fflush(stderr);
}

/**
 *  A signal was caught, siglongjmp called and main() has called this.
 *  We do not deallocate stuff so it can be found in the core dump.
 *
 *  @param[in] sig the signal number
 */
static _Noreturn void
cleanup_and_abort(int sig)
{
    if (processing_state == PROC_STATE_INIT) {
        fprintf(stderr, AG_NEVER_STARTED, sig, strsignal(sig));
        exit(AUTOGEN_EXIT_SIGNAL + sig);
    }

    if (*oops_pfx != NUL) {
        fputs(oops_pfx, stderr);
        oops_pfx = zNil;
    }

    fprintf(stderr, AG_SIG_ABORT_FMT, sig, strsignal(sig),
            ((unsigned)processing_state <= PROC_STATE_DONE)
            ? state_names[ processing_state ] : BOGUS_TAG);

    if (processing_state == PROC_STATE_ABORTING) {
        exit_cleanup(EXIT_PCLOSE_NOWAIT);
        abort();
    }

    processing_state = PROC_STATE_ABORTING;
    setup_signals(SIG_DFL, SIG_DFL, SIG_DFL);

    /*
     *  IF there is a current template, then we should report where we are
     *  so that the template writer knows where to look for their problem.
     */
    if (current_tpl != NULL) {
        int line;
        mac_func_t fnCd;
        char const * pzFn;
        char const * pzFl;

        if (cur_macro == NULL) {
            pzFn = PSEUDO_MACRO_NAME_STR;
            line = 0;
            fnCd = (mac_func_t)-1;

        } else {
            mac_func_t f =
                (cur_macro->md_code > FUNC_CT)
                    ? FTYP_SELECT : cur_macro->md_code;
            pzFn = ag_fun_names[ f ];
            line = cur_macro->md_line;
            fnCd = cur_macro->md_code;
        }
        pzFl = current_tpl->td_file;
        if (pzFl == NULL)
            pzFl = NULL_FILE_NAME_STR;
        fprintf(stderr, AG_ABORT_LOC_FMT, pzFl, line, pzFn, fnCd);
    }

#ifdef HAVE_SYS_RESOURCE_H
    /*
     *  If `--core' has been specified and if we have get/set rlimit,
     *  then try to set the core limit to the "hard" maximum before aborting.
     */
    if (HAVE_OPT(CORE)) {
        struct rlimit rlim;
        getrlimit(RLIMIT_CORE, &rlim);
        rlim.rlim_cur = rlim.rlim_max;
        setrlimit(RLIMIT_CORE, &rlim);
    }
#endif

    exit_cleanup(EXIT_PCLOSE_NOWAIT);
    abort();
}


/**
 *  catch_sig_and_bail catches signals we abend on.  The "siglongjmp"
 *  goes back to the real "main()" procedure and it will call
 *  "cleanup_and_abort()", above.
 *
 *  @param[in] sig the signal number
 */
static void
catch_sig_and_bail(int sig)
{
    switch (processing_state) {
    case PROC_STATE_ABORTING:
        ag_exit_code = AUTOGEN_EXIT_SIGNAL + sig;

    case PROC_STATE_DONE:
        break;

    default:
        cleanup_and_abort(sig);
    }
}


/**
 *  ignore_signal is the handler for SIGCHLD.  If we set it to default,
 *  it will kill us.  If we set it to ignore, it will be inherited.
 *  Therefore, always in all programs set it to call a procedure.
 *  The "wait(3)" call will do magical things, but will not override SIGIGN.
 *
 *  @param[in] sig the signal number
 */
static void
ignore_signal(int sig)
{
    (void)sig;
    return;
}


/**
 *  This is _always_ called for exit processing.
 *  This only returns if "exit(3C)" is called during option processing.
 *  We have no way of determining the correct exit code.
 *  (Requested version or help exits EXIT_SUCCESS.  Option failures
 *  are EXIT_FAILURE.  We cannot determine here.)
 */
static void
done_check(void)
{
    /*
     *  There are contexts wherein this function can get called twice.
     */
    {
        static int done_check_done = 0;
        if (done_check_done) {
            if (OPT_VALUE_TRACE > TRACE_NOTHING)
                fprintf(trace_fp, DONE_CHECK_REDONE);
            return;
        }
        done_check_done = 1;
    }

    switch (processing_state) {
    case PROC_STATE_EMITTING:
    case PROC_STATE_INCLUDING:
        /*
         * A library (viz., Guile) procedure has called exit(3C).  The
         * AutoGen abort paths all set processing_state to PROC_STATE_ABORTING.
         */
        if (*oops_pfx != NUL) {
            /*
             *  Emit the CGI page header for an error message.  We will rewind
             *  stderr and write the contents to stdout momentarily.
             */
            fputs(oops_pfx, stdout);
            oops_pfx = zNil;
        }

        fprintf(stderr, SCHEME_EVAL_ERR_FMT, current_tpl->td_file,
                cur_macro->md_line);

        /*
         *  We got here because someone called exit early.
         */
        do  {
#ifndef DEBUG_ENABLED
            out_close(false);
#else
            out_close(true);
#endif
        } while (cur_fpstack->stk_prev != NULL);
        ag_exit_code = AUTOGEN_EXIT_BAD_TEMPLATE;
        break; /* continue failure exit */

    case PROC_STATE_LOAD_DEFS:
        ag_exit_code = AUTOGEN_EXIT_BAD_DEFINITIONS;
        /* FALLTHROUGH */

    default:
        fprintf(stderr, AG_ABEND_STATE_FMT, state_names[processing_state]);
        goto autogen_aborts;

    case PROC_STATE_ABORTING:
        ag_exit_code = AUTOGEN_EXIT_BAD_TEMPLATE;

    autogen_aborts:
        if (*oops_pfx != NUL) {
            /*
             *  Emit the CGI page header for an error message.  We will rewind
             *  stderr and write the contents to stdout momentarily.
             */
            fputs(oops_pfx, stdout);
            oops_pfx = zNil;
        }
        break; /* continue failure exit */

    case PROC_STATE_OPTIONS:
        /* Exiting in option processing state is verbose enough */
        break;

    case PROC_STATE_DONE:
        break; /* continue normal exit */
    }

    if (last_scm_cmd != NULL)
        fprintf(stderr, GUILE_CMD_FAIL_FMT, last_scm_cmd);

    /*
     *  IF we diverted stderr, then now is the time to copy the text to stdout.
     *  This is done for CGI mode wherein we produce an error page in case of
     *  an error, but otherwise discard stderr.
     */
    if (cgi_stderr != NULL) {
        do {
            long pos = ftell(stderr);
            char * pz;
            size_t rdlen;

            /*
             *  Don't bother with the overhead if there is no work to do.
             */
            if (pos <= 0)
                break;
            pz = AGALOC(pos, "stderr redir");
            rewind(stderr);
            rdlen = fread( pz, (size_t)1, (size_t)pos, stderr);
            fwrite(pz, (size_t)1, rdlen, stdout);
            AGFREE(pz);
        } while (false);

        unlink(cgi_stderr);
        AGFREE(cgi_stderr);
        cgi_stderr = NULL;
    }

    scribble_deinit();

    if (OPT_VALUE_TRACE > TRACE_NOTHING)
        fprintf(trace_fp, DONE_CHECK_DONE);

    exit_cleanup(EXIT_PCLOSE_WAIT);

    /*
     *  When processing options, return so that the option processing exit code
     *  is used.  Otherwise, terminate execution now with the decided upon
     *  exit code.  (Always EXIT_FAILURE, unless this was called at the end
     *  of inner_main().)
     */
    if (processing_state != PROC_STATE_OPTIONS)
        exit(ag_exit_code);
}


LOCAL _Noreturn void
ag_abend_at(char const * msg
#ifdef DEBUG_ENABLED
            , char const * fname, int line
#endif
    )
{
    if (*oops_pfx != NUL) {
        fputs(oops_pfx, stderr);
        oops_pfx = zNil;
    }

#ifdef DEBUG_ENABLED
    fprintf(stderr, "Giving up in %s line %d\n", fname, line);
#endif

    if ((processing_state >= PROC_STATE_LIB_LOAD) && (current_tpl != NULL)) {
        int l_no = (cur_macro == NULL) ? -1 : cur_macro->md_line;
        fprintf(stderr, ERROR_IN_TPL_FMT, current_tpl->td_file, l_no);
    }
    fputs(msg, stderr);
    msg += strlen(msg);
    if (msg[-1] != NL)
        fputc(NL, stderr);

#ifdef DEBUG_ENABLED
    abort();
#else
    {
        proc_state_t o_state = processing_state;
        processing_state = PROC_STATE_ABORTING;

        switch (o_state) {
        case PROC_STATE_EMITTING:
        case PROC_STATE_INCLUDING:
        case PROC_STATE_CLEANUP:
            longjmp(abort_jmp_buf, FAILURE);
            /* NOTREACHED */
        default:
            abort();
            /* NOTREACHED */
        }
    }
#endif
}

static void
setup_signals(sighandler_proc_t * hdl_chld,
              sighandler_proc_t * hdl_abrt,
              sighandler_proc_t * hdl_dflt)
{
    struct sigaction  sa;
    int    sigNo  = 1;
    int const maxSig = MAX_AG_SIG;

    sa.sa_flags   = 0;
    sigemptyset(&sa.sa_mask);

    do  {
        switch (sigNo) {
            /*
             *  Signal handling for SIGCHLD needs to be ignored.  Do *NOT* use
             *  SIG_IGN to do this.  That gets inherited by programs that need
             *  to be able to use wait(2) calls.  At the same time, we don't
             *  want our children to become zombies.  We may run out of zombie
             *  space.  Therefore, set the handler to an empty procedure.
             *  POSIX oversight.  Allowed to be fixed for next POSIX rev, tho
             *  it is "optional" to reset SIGCHLD on exec(2).
             */
        case SIGCHLD:
            sa.sa_handler = hdl_chld;
            break;

        case SIGABRT:
            sa.sa_handler = hdl_abrt;
            break;

            /*
             *  Signals we must leave alone.
             */
        case SIGKILL:
        case SIGSTOP:

            /*
             *  Signals we choose to leave alone.
             */
#ifdef SIGTSTP
        case SIGTSTP:
#endif
#ifdef SIGPROF
        case SIGPROF:
#endif
            continue;

#if defined(DEBUG_ENABLED)
        case SIGBUS:
        case SIGSEGV:
            /*
             *  While debugging AutoGen, we want seg faults to happen and
             *  trigger core dumps.  Make sure this happens.
             */
            sa.sa_handler = SIG_DFL;
            break;
#endif

            /*
             *  Signals to ignore with SIG_IGN.
             */
        case 0: /* cannot happen, but the following might not be defined */
#ifdef SIGWINCH
        case SIGWINCH:
#endif
#ifdef SIGTTIN
        case SIGTTIN:  /* tty input  */
#endif
#ifdef SIGTTOU
        case SIGTTOU:  /* tty output */
#endif
            sa.sa_handler = SIG_IGN;
            break;

        default:
            sa.sa_handler = hdl_dflt;
        }
        sigaction(sigNo,  &sa, NULL);
    } while (++sigNo < maxSig);
}

#ifndef HAVE_STRFTIME
#  include <compat/strftime.c>
#endif

#ifndef HAVE_STRSIGNAL
#  include <compat/strsignal.c>
#endif

LOCAL void *
ao_malloc (size_t sz)
{
    void * res = malloc(sz);
    if (res == NULL)
        die(AUTOGEN_EXIT_FS_ERROR, MALLOC_FAIL_FMT, sz);
    return res;
}


LOCAL void *
ao_realloc (void *p, size_t sz)
{
    void * res = (p == NULL) ? malloc(sz) : realloc(p, sz);
    if (res == NULL)
        die(AUTOGEN_EXIT_FS_ERROR, REALLOC_FAIL_FMT, sz, p);
    return res;
}

LOCAL char *
ao_strdup (char const * str)
{
    char * res = strdup(str);
    if (res == NULL)
        die(AUTOGEN_EXIT_FS_ERROR, STRDUP_FAIL_FMT, (int)strlen(str));
    return res;
}

#ifdef __GNUC__
    void ignore_vars(void);
    void ignore_vars(void) {
        (void)option_load_mode, (void)program_pkgdatadir;
    }
#endif
/**
 * @}
 *
 * Local Variables:
 * mode: C
 * c-file-style: "stroustrup"
 * indent-tabs-mode: nil
 * End:
 * end of agen5/autogen.c */
